aten/src/ATen/native/cuda/UpSampleNearest2d.cu

```diff
@@ -11,110 +11,114 @@ namespace at {
+  const size_t w1 = width1 == width2
+      ? w2
+      : nearest_neighbor_compute_source_index(width_scale, w2, width1);
+
+  size_t src_index = (nc_iter * height1 + h1) * width1 + w1;
+  size_t src_index_stride = nc_stride * width1 * height1;
+  size_t dst_index = (nc_iter * height2 + h2) * width2 + w2;
+  size_t dst_index_stride = nc_stride * width2 * height2;
+
+  // iterating over
+  while (nc_iter < nc) {
+    odata[dst_index] = idata[src_index];
+    dst_index += dst_index_stride;
+    src_index += src_index_stride;
+    nc_iter += nc_stride;
   }
 }
 
-// Backward operation
 template <typename scalar_t, typename accscalar_t>
 C10_LAUNCH_BOUNDS_1(1024)
 __global__ void upsample_nearest2d_backward_out_frame(
-    const int n,
-    PackedTensorAccessor<scalar_t, 4> idata,
-    const PackedTensorAccessor<scalar_t, 4> odata) {
-  int index = threadIdx.x + blockIdx.x * blockDim.x;
-
-  const int batchsize = idata.size(0);
-  const int channels = idata.size(1);
-  const int height1 = idata.size(2);
-  const int width1 = idata.size(3);
-  const int height2 = odata.size(2);
-  const int width2 = odata.size(3);
-
-  const float height_scale = (float)height1 / (float)height2;
-  const float width_scale = (float)width1 / (float)width2;
-
-  if (index < n) {
-    const int w2 = index % width2; // 0:width2-1
-    const int h2 = index / width2; // 0:height2-1
-    // special case: just copy
-    if (height1 == height2 && width1 == width2) {
-      const int h1 = h2;
-      const int w1 = w2;
-
-      for (int n = 0; n < batchsize; n++) {
-        for (int c = 0; c < channels; ++c) {
-          const scalar_t val = odata[n][c][h2][w2];
-          idata[n][c][h1][w1] = val;
-        }
-      }
-      return;
-    }
-    //
-    const int h1 =
-        nearest_neighbor_compute_source_index(height_scale, h2, height1);
-    const int w1 =
-        nearest_neighbor_compute_source_index(width_scale, w2, width1);
-
-    for (int n = 0; n < batchsize; n++) {
-      for (int c = 0; c < channels; ++c) {
-        const scalar_t d2val = odata[n][c][h2][w2];
-        atomicAdd(&idata[n][c][h1][w1], d2val);
+    const scalar_t* grad_o,
+    size_t dim_b,
+    size_t dim_c,
+    size_t src_dim_h,
+    size_t src_dim_w,
+    size_t dst_dim_h,
+    size_t dst_dim_w,
+    scalar_t* grad_i) {
+  assert(gridDim.y == 1);
+  assert(gridDim.z == 1);
+  assert(blockDim.y == 1);
+  assert(blockDim.z == 1);
+
+  size_t dst_idx = blockIdx.x * blockDim.x + threadIdx.x;
+  if (dst_idx >= dim_c * dst_dim_h * dst_dim_w)
+    return;
+
+  int dst_c_stride = dst_dim_h * dst_dim_w;
+  int src_c_stride = src_dim_h * src_dim_w;
+
+  int c = (dst_idx / (dst_c_stride)) % dim_c;
+
+  float scale_factor = (float)src_dim_h / dst_dim_h;
+  int dst_y = (dst_idx / dst_dim_w) % dst_dim_h;
+  int src_y = (int)ceilf(dst_y * scale_factor);
+  int src_y_up = min((int)ceilf((dst_y + 1) * scale_factor), (int)src_dim_h);
+
+  scale_factor = (float)src_dim_w / dst_dim_w;
```

Comment:
Sometimes you cast one operand, sometimes you cast both - is there any reason to it?

Reply:
I think Alexey's kernel casts only one operand, while the upstream code casts both. So they were sorta consistent in their own ways.

I don't think there's any perf impact on this, as they would have been casted to float any way.
I'll update to cast both just for consistency with upstream code.